content/posts/connecting-go-profiling-with-tracing/index.md

```diff
@@ -0,0 +1,132 @@
+---
+title: "Connecting Go Profiling With Tracing"
+date: 2022-02-07T10:00:00+01:00
+slug: "connecting-go-profiling-with-tracing"
+---
+
+Today I'd like to share a few thoughts on a few new [profiling features](https://docs.datadoghq.com/tracing/profiler/connect_traces_and_profiles/) we recently shipped at Datadog. I'll explain what they do, how they work, and which Go 1.18 contributions were needed in order for things to work well.
+
+## Feature Showcase
+
+Let's start with the end result. Imagine you have a 100ms trace where 10ms are spend on a database query, but 90ms remain unexplained.
+
+<img width="100%" src="./flamechart.png"/>
+
+When this happens, you usually need to study your code for clues. Did you forget some tracing instrumentation? Time add it, redeploy and wait. Or perhaps you need to optimize your Go code? If yes, how?
+
+This workflow is manageable, but it turns out there is a better way – we can use profiling data to fill in the gaps. And that's exactly what our new [Code Hotspots](https://docs.datadoghq.com/tracing/profiler/connect_traces_and_profiles/#identify-code-hotspots-in-slow-traces) feature does. As you can see below, our request used 90ms On-CPU time. This is a strong signal that lets us rule out Off-CPU activity such as uninstrumented service calls, mutex contentions, channel waits, sleeping, etc.
+
+<img width="100%" src="./hotspots.png"/>
+
+Even better, when clicking the "View Profile" button, we can view this On-CPU time as a per-request flame graph. Here we can see that our time was spent on JSON encoding.
+
+And since our http handler functions don't show up in the stack traces, we can also indirectly infer that this work was done in a background goroutine spawned by the goroutine handling the request.
```

Comment:
```suggestion
And since our HTTP handler functions don't show up in the stack traces, we can also indirectly infer that this work was done in a background goroutine spawned by the goroutine handling the request.
```